test/unit/modules/blocks/chain.js

```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
+						{ id: 6, height: 6, transactions: undefined },
+						dbStub.tx
+					)
+					.catch(err => {
+						expect(err.message).to.equal(
+							'expecting an array or an iterable object but got [object Null]'
+						);
+						done();
+					});
 			});
+		});
 
-			describe('when modules.accounts.setAccountAndGet succeeds', () => {
-				it('should call __private.applyTransaction');
+		describe('when block transaction is empty', () => {
+			it('should return resolved promise with no error', done => {
+				__private
+					.applyConfirmedStep({ id: 6, height: 6, transactions: [] }, dbStub.tx)
+					.then(resolved => {
+						expect(resolved).to.be.an('array').that.is.empty;
+						done();
+					});
+			});
+		});
 
-				it('should call __private.applyTransaction with block');
+		describe('when block.transaction is not empty', () => {
+			describe('modules.accounts.getAccount', () => {
+				describe('when fails', () => {
+					beforeEach(() => {
+						return modules.accounts.getAccount.callsArgWith(
+							1,
+							'getAccount-ERR',
+							null
+						);
+					});
 
-				it('should call __private.applyTransaction with transaction');
+					afterEach(() => {
+						expect(modules.accounts.getAccount.callCount).to.equal(1);
+						expect(modules.transactions.apply.callCount).to.equal(0);
+						expect(loggerStub.error.args[0][0]).to.equal(
+							'Failed to get account to apply transaction: 1 - getAccount-ERR'
+						);
+						expect(loggerStub.error.args[1][0]).to.equal('Transaction');
+						return expect(loggerStub.error.args[1][1]).to.deep.equal({
+							id: 1,
+							type: 0,
+							senderPublicKey: 'a',
+						});
+					});
 
-				it('should call __private.applyTransaction with sender');
+					it('should return rejected promise with error', done => {
+						__private
+							.applyConfirmedStep(
+								{
+									id: 5,
+									height: 5,
+									transactions: [
+										{ id: 1, type: 0, senderPublicKey: 'a' },
+										{ id: 2, type: 1, senderPublicKey: 'b' },
+									],
+								},
+								dbStub.tx
+							)
+							.catch(err => {
+								expect(err).to.equal(
+									'Failed to get account to apply transaction: 1 - getAccount-ERR'
+								);
+								done();
+							});
+					});
+				});
 
-				it('should call __private.applyTransaction with callback');
+				describe('when succeeds', () => {
+					beforeEach(() => {
+						return modules.accounts.getAccount.callsArgWith(1, null, 'sender1');
+					});
 
-				it('should call tracker.applyNext');
-			});
+					describe('library.logic.transaction.apply', () => {
+						describe('when fails', () => {
+							beforeEach(() => {
+								return modules.transactions.apply.callsArgWith(
+									3,
+									'apply-ERR',
+									null
+								);
+							});
 
-			describe('after loop through block.transactions', () => {
-				it('should set genesis block as last block');
+							afterEach(() => {
+								expect(modules.accounts.getAccount.callCount).to.equal(1);
```

Comment:
i don't like these assertions in afterEach. If we want to have something like this, we can create a function, and call it from all tests.